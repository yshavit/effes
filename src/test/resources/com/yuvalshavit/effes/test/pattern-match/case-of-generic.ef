type Empty
type Node[T](head: T, tail: Node[T] | Empty)

type TrueNode
type FalseNode
type Boolean = True | False

describe (in: Node[True] | Node[False] | Boolean) -> TrueNode | FalseNode | Boolean:
  return case in of
    Node[True](head, tail): TrueNode
    Node[False](head, tail): FalseNode
    True: True
    False: False

main:
  print describe Node[True](True, Empty)
  print describe Node[False](False, Empty)
  print describe True
