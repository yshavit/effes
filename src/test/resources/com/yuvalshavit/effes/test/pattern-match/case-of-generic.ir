def describe (in: False | Node[False] | Node[True] | True) -> False | FalseNode | True | TrueNode:
  ReturnStatement
    CaseExpression
      case:
        VarExpression
          arg: in
          pos: 0
      of Node[T](True, tail):
        CtorInvoke
          type: TrueNode
      of Node[T](False, tail):
        CtorInvoke
          type: FalseNode
      of True:
        CtorInvoke
          type: True
      of False:
        CtorInvoke
          type: False
