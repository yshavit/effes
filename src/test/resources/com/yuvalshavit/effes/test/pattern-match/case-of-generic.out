Your argument was: TrueNode
Your argument was: FalseNode
Your argument was: True
>> null
